### PR TITLE
Implement basic app.ini and path checks to doctor cmd

### PR DESCRIPTION
**Pull request by guillep2k**
This PR adds some basic `app.ini` and path checking to the `doctor` command.

In particular, it won't proceed if `app.ini` doesn't look healthy enough, and will provide information about the application paths the user may find useful.

### Normal run
```
$ gitea doctor
[ 1 ] Check paths and basic configuration
- Configuration File Path:   '/home/gittest/etc/app.ini'
- Repository Root Path:      '/home/gittest/gitea-repositories'
- Data Root Path:            '/home/gittest/data'
- Custom File Root Path:     '/home/gittest/custom'
- Work directory:            '/home/gittest'
- Log Root Path:             '/home/gittest/log'
OK.

[ 2 ] Check if OpenSSH authorized_keys file id correct
OK.
```

### Run with an empty `app.ini`
```
$ gitea --config `pwd`/empty.app.ini doctor
[ 1 ] Check paths and basic configuration
- Configuration File Path:   '/home/gittest/tmp/empty.app.ini'
- Repository Root Path:      '/home/gittest/gitea-repositories'
- Data Root Path:            '/home/gittest/tmp/data'
-     ERROR: not writable: open /home/gittest/tmp/data/doctors-order772291956: permission denied
- Custom File Root Path:     '/home/gittest/tmp/custom'
-     NOTICE: not accessible (stat /home/gittest/tmp/custom: no such file or directory)
- Work directory:            '/home/gittest/tmp'
- Log Root Path:             '/home/gittest/tmp/log'
-     ERROR: stat /home/gittest/tmp/log: no such file or directory
Error: Please check your configuration file and try again.
```
### Run with no `app.ini` at all

```
$ gitea doctor
[ 1 ] Check paths and basic configuration
- Failed to find configuration file at '/home/gittest/tmp/custom/conf/app.ini'.
- If you've never ran Gitea yet, this is normal and '/home/gittest/tmp/custom/conf/app.ini' will be created for you.
- Otherwise check that you are running this command from the correct path and/or provide the `--config` option.
Error: Can't proceed without a configuration file
```

Note: I've "silenced" the default logger since its output (being asynchronous) was mixing in between the doctor's output and I find it confusing:

```
$ gitea doctor
[ 1 ] Check paths and basic configuration
- Configuration File Path:   '/home/gittest/etc/app.ini'
- Repository Root Path:      '/home/gittest/gitea-repositories'
2020/01/29 15:52:04 ...dules/setting/git.go:91:newGit() [I] Git Version: 2.22.0, Wire Protocol Version 2 Enabled
- Data Root Path:            '/home/gittest/data'
- Custom File Root Path:     '/home/gittest/custom'
- Work directory:            '/home/gittest'
- Log Root Path:             '/home/gittest/log'
2020/01/29 15:52:04 ...dules/setting/git.go:91:newGit() [I] Git Version: 2.22.0, Wire Protocol Version 2 Enabled
2020/01/29 15:52:04 ...dules/setting/log.go:185:generateNamedLogger() [I] Xorm Log: File(file:trace)
2020/01/29 15:52:04 ...dules/setting/log.go:185:generateNamedLogger() [I] Xorm Log: Console(console:error)
OK.

[ 2 ] Check if OpenSSH authorized_keys file id correct
OK.
```

I think it will be best to create a separate log for all of doctor's activities, but I leave that for another PR.

### CODE REVIEW

**Comment by 6543** [comment]
O I had a issue once a time:
instead of `;` i had a `:` as comment indicator -> witch result in a gitea crash without any information left where I could find the issue ...

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
@guillep2k is there a chance to add `:` config checker?

File: cmd/doctor.go [hunk 1]
```diff
@@ -74,6 +92,80 @@ func exePath() (string, error) {
...
+			}
+		}
+		return
+	}
+
+	// Note print paths inside quotes to make any leading/trailing spaces evident
+	check("Configuration File Path", setting.CustomConf, false, false)
+	check("Repository Root Path", setting.RepoRootPath, true, true)
+	check("Data Root Path", setting.AppDataPath, true, true)
+	check("Custom File Root Path", setting.CustomPath, true, false)
```
Comment (jolheiser) [inline cmd/doctor.go]: Since a `custom` folder isn't required to run Gitea (at least not that I recall), the message "ERROR" may be a little misleading.

Comment (guillep2k) [reply cmd/doctor.go]: @jolheiser True that. I've changed it into `NOTICE: not accessible (msg)`.

File: cmd/doctor.go [hunk 2]
```diff
@@ -22,18 +25,24 @@ import (
 // CmdDoctor represents the available doctor sub-command.
 var CmdDoctor = cli.Command{
 	Name:        "doctor",
-	Usage:       "Diagnose the problems",
-	Description: "A command to diagnose the problems of current gitea instance according the given configuration.",
+	Usage:       "Diagnose and repair problems",
+	Description: "A command to diagnose and repair problems of current gitea instance according the given configuration.",
```
Comment (jolheiser) [inline cmd/doctor.go]: Does this command intend to actually _repair_ problems or just report them to the user?

**Comment by guillep2k** [comment]
> 
> 
> @guillep2k is there a chance to add `:` config checker?

Maybe next time. I'd need to investigate all the possible valid syntaxes for `app.ini` lines and make a regex for that. It might not be a trivial task.

File: cmd/doctor.go [hunk 3]
```diff
@@ -42,21 +51,30 @@ var checklist = []check{
...
-		return nil
-	}
+
+	// Silence the console logger
+	// TODO: Redirect all logs into `doctor.log` ignoring any other log configuration
+	log.DelNamedLogger("console")
+	log.DelNamedLogger(log.DEFAULT)
 
 	for i, check := range checklist {
+		if i == 1 {
```
Comment (jolheiser) [inline cmd/doctor.go]: Just a passing comment, "magic" indexes like this are sometimes missed. I don't have a real suggestion at the moment, just wanted to make note of it.

File: cmd/doctor.go [hunk 4]
```diff
@@ -22,18 +25,24 @@ import (
...
-	Description: "A command to diagnose the problems of current gitea instance according the given configuration.",
+	Usage:       "Diagnose and repair problems",
+	Description: "A command to diagnose and repair problems of current gitea instance according the given configuration.",
 	Action:      runDoctor,
 }
 
 type check struct {
 	title string
 	f     func(ctx *cli.Context) ([]string, error)
+	abort bool
```
Comment (jolheiser) [inline cmd/doctor.go]: Can this variable be renamed to indicate it is to abort if there are errors with a particular check?  
`abortIfError` doesn't look very pretty but makes more sense to me.

File: cmd/doctor.go (hunk 2 above)
Comment (guillep2k) [reply cmd/doctor.go]: _Repairing_ was my intention... but I didn't get too far. 😅
Done.

File: cmd/doctor.go (hunk 4 above)
Comment (guillep2k) [reply cmd/doctor.go]: > 
> 
> Can this variable be renamed to indicate it is to abort if there are errors with a particular check?
> `abortIfError` doesn't look very pretty but makes more sense to me.

Done.

File: cmd/doctor.go (hunk 3 above)
Comment (guillep2k) [reply cmd/doctor.go]: > 
> 
> Just a passing comment, "magic" indexes like this are sometimes missed.

Done.

File: cmd/doctor.go [hunk 5]
```diff
@@ -22,18 +25,27 @@ import (
 // CmdDoctor represents the available doctor sub-command.
 var CmdDoctor = cli.Command{
 	Name:        "doctor",
-	Usage:       "Diagnose the problems",
-	Description: "A command to diagnose the problems of current gitea instance according the given configuration.",
+	Usage:       "Diagnose problems",
+	Description: "A command to diagnose problems of current gitea instance according the given configuration.",
```
Comment (jolheiser) [inline cmd/doctor.go]:
```suggestion
	Description: "A command to diagnose problems with the current Gitea instance according to the given configuration.",
```

**Review by jolheiser — APPROVED** [review]
One suggestion, otherwise LGTM

File: modules/options/static.go [hunk 6]
```diff
@@ -112,3 +112,7 @@ func fileFromDir(name string) ([]byte, error) {
 
 	return ioutil.ReadAll(f)
 }
+
+func IsDynamic() bool {
```
Comment (guillep2k) [inline modules/options/static.go:117]:
```suggestion
// IsDynamic will return false when using embedded data (-tags bindata)
func IsDynamic() bool {
```

File: modules/options/dynamic.go [hunk 7]
```diff
@@ -98,3 +98,7 @@ func fileFromDir(name string) ([]byte, error) {
 
 	return []byte{}, fmt.Errorf("Asset file does not exist: %s", name)
 }
+
+func IsDynamic() bool {
```
Comment (guillep2k) [inline modules/options/dynamic.go:103]:
```suggestion
// IsDynamic will return false when using embedded data (-tags bindata)
func IsDynamic() bool {
```

**Comment by guillep2k** [comment]
Related: https://github.com/go-gitea/gitea/issues/8781#issuecomment-579743599

**Pull request merged**